### DISCUSSION TITLE
How to measure position of motor

### DISCUSSION THREAD

**Discussion by GenghisHan-op** [pybricks/support#1913] [General]
<img width="634" alt="Screenshot 2024-11-03 175343" src="https://github.com/user-attachments/assets/8d500a85-851e-4ced-a47f-3edef94bdacb">
I have a Motor connect to a gear like in the image below, controlled by an Xbox controller, please teach me how can I make it run within the upper and lower limit only, if it reach the limits, the motor stop. Starting position of the gear is always at lower limit. 
I tried using control.target_tolerances()→ Tuple[[int], [int]] as instruct from the document but no luck.
Sorry for my bad English.
Thanks in advance.

**Comment by BertLindeman** [comment]
Hello,

Do you have an example of what you tried and tell (show) us how that fails?
If you show us the program it is easier to point you to improvements.

Bert

**Reply by GenghisHan-op** [reply]
I'm just trying to get the position in mm using control.target_tolerances() but it seems that I cannot make it to work, I cannot debug because the hub disconnected from PC the moment it connect to the controller.

**Reply by BertLindeman** [reply]
Ah Technichub and xbox controller. 

That is hard to debug for sure.
You could try a small test program without the xbox controller so you can see what happens.

The calculation is (hope it is correct) 360 degrees rotation of the motor is 8 tooth horizontal distance. 
(I assume the 8 tooth gear). 
You can count how many teeth there are on the straight gear. 
Then you can divide that length by 8.
The angle of the motor you can determine with ```motor.angle()```  

Would this help you?

Bert

**Reply by GenghisHan-op** [reply]
Thanks for your reply, 
Your math is exactly what I started, but the problem is the motor could stop in the middle by the control of controller, not just go straight from lower to upper limit.
I think the angle will reset each time the motor start moving again? Is that right?
Thanks again for your help

**Reply by BertLindeman** [reply]
Assume the type of motor you use is a motor with rotation sensor like they are shown in the pybricks documentation.
 
The motor remembers its angle, even if not connected or hub is powered off.
A simple program to verify:
```python
from pybricks.pupdevices import Motor
from pybricks.parameters import Port

motor = Motor(Port.A)
print("start at", motor.angle())
```

I connected a random motor (that happened to be next to me and ran this program twice.
The result: 
```
start at -51
start at -51
```

**Reply by GenghisHan-op** [reply]
<img width="513" alt="image" src="https://github.com/user-attachments/assets/81cbc5e8-9980-478d-891d-27395bcb4144">
I tried this code, as you can see, the last angle of the motor before i stop the program is 191, then I started the motor the second time, and the first angle is -32.
I thought it should be 191.

**Reply by GenghisHan-op** [reply]
Oh nevermind, I changed reset_angle to False and everything works like a charm.
Problem solved.
Thanks for all your help

**Reply by BertLindeman** [reply]
Glad you solved it.
Maybe show us later what you made? 

Happy pybricks coding,

Bert

**Reply by GenghisHan-op** [reply]
```
motor = Motor(Port.A, Direction.CLOCKWISE, [8, 26],False)
speed = controller.joystick_right()[1]
if speed != 0:
        if speed < 0:
            motor.dc(speed)
            angle = motor.angle()
            if angle < -8600:
                motor.dc(0)              
        if speed > 0:
            motor.dc(speed)
            angle = motor.angle()
            if angle > 0:
                motor.dc(0)
else:
        motor.dc(0)
```

This is my code. The limit is 0 and 8600 (angle). The motor is controlled by the controller, with this code, the gear never move pass the 2 limits. If you can improve, please show me. Thank you.

**Reply by BertLindeman** [reply]
I see no loop in this snippet. If this is the complete test program, then this is all done only once and the program is ready in no time.

After the define of the motor you might need something like:
```python
while True:
    # do the speed control like you show
    wait(100)  # just a small wait
```

Bert

**Reply by GenghisHan-op** [reply]
Ah, this is not a complete program, just a chunk of code that do the work in the first post.

**Reply by BertLindeman** [reply]
I simplified the snippet a bit, but kept the logic:
```python

while True:
    speed = controller.joystick_right()[1]
    if speed < 0:
        if motor.angle() < -8600:
            motor.dc(0)
        else:
            motor.dc(speed)
    elif speed > 0:
        if motor.angle() > 0:
            motor.dc(0)
        else:
            motor.dc(speed)
    else:
        motor.dc(0)
```

And this works ok on a primehub, so I can see some debugging.
Might there be some error in your complete program?
Due to the disconnect on the TechnicHub for the xbox controller you cannot see the runtime error.

[EDIT] You might simulate (if you have a port free) the joystick with a color sensor. Say RED makes speed -100 and GREEN does +100. By running without xbox controller you may find out where the problem is.

[EDIT 2] The motor has inertia, so stopping at -8600 will have the motor stopped at (say) -8634 or so.
Keep the deceleration in mind.

**Reply by GenghisHan-op** [reply]
Thanks for your reply, the code work without any problem on my technic hub. Maybe I'll change to primehub later. The error ratio of the motor is small and acceptable. I do have color sensor but I have no room to install it, the technic hub still have 1 port free.
